### DISCUSSION TITLE
How to get bounce data?

### DISCUSSION THREAD

**Discussion by kayastha1988** [postalserver/postal#2464] [Development discussions]
After running a email campaign, how to grab the bounce record data so that I can managed them in API and send that data to third party email server provider. Is there any way to grab those bounce email data records???
Thanks

**Comment by sinnbeck** [comment]
You need to set up webhooks to send that data to your website that sent the emails. https://docs.postalserver.io/developer/webhooks

Be aware that you need to set this up before sending the messages

**Reply by kayastha1988** [reply]
I have setup this webhooks but I am confused how to grab those bounce records. I can only see those information in server only.

**Reply by sinnbeck** [reply]
Did you write code on the website part to handle the incoming webhooks? What is your backend coded in?

**Reply by kayastha1988** [reply]
Actually we have 3rd party service called OEMPRO to send email campaign just like sendgrid or other email marketing tools. For SMTP server, I am using postal as it's free and opensource program. From dashboard portal, I have done some work for webhooks. Is there anything that we can do go create API or some json file or anything to grab the bounce records data from postal smtp server?

**Reply by sinnbeck** [reply]
There is an API, but it is quite limited https://docs.postalserver.io/developer/api https://apiv1.postalserver.io/controllers/messages/message.html

I would highly recommend using webhooks instead. It seems your system supports incoming webhooks: https://help.octeth.com/webhooks (I assume this is it): If that cannot do what you need, then you might instead need to set up your own external system for handling the webhooks :)

**Reply by kayastha1988** [reply]
Yes, it is. Thanks for  help. I will have a look on this webhook docs. Thanks

**Reply by kayastha1988** [reply]
Hello @sinnbeck One quick question. **Octeth** will give us a webhook url and that url is enough to get all the bounce records from postal. Is that correct? I have already created webhooks in postal server but confused what url should be defined. In postal, there is place to define url for webhooks and webhooks url from octeth contains email parameter as well. How that param can be handled?

Looking forward to hearing from you.

Thanks

**Reply by sinnbeck** [reply]
I dont know anything about Octeth sadly and their docs are very sparse. But you need to set up and url inside of that software that you can put inside the webhooks section of postal. Then inside Octeth you need to tell it to check if the event type is `MessageBounced` and then match the message using the `original_message.message_id` https://docs.postalserver.io/developer/webhooks/#message-bounces

There is a chance this isnt possible at all, in which case you have 2 options.
1. Build your own webhooks handler and put it on a server, as a middleman (I wrote mine in php)
2. Wait until the next major release of the API for postal. I believe there is a coming version with more features somewhere in the future

**Reply by kayastha1988** [reply]
**Build your own webhooks handler and put it on a server, as a middleman (I wrote mine in php)**

This is webhooks that will grab all the bounce data from postal which will be used to send data to octeth where they need to check if the event as you said above. Is that correct?
Would you mind if I can have a look to the example code, if that's possible?

Thanks

**Reply by sinnbeck** [reply]
Let me quickly explain what a webhook is just so you understand better. A webhook is a http request that postal will send over to some external url. This url will then accept the request and do something with the data.

My code is something like this
```
$payload = $this->webhookCall->payload;
match ($payload['event']) {
            'MessageSent' => $this->handleSent()
            'MessageBounced' => $this->handleBounced(),
            'MessageDelayed' => $this->handleDelayed(),
            'MessageDeliveryFailed' => $this->handleFailed(),
            'MessageHeld' => $this->handleHeld(),
        };

**Reply by kayastha1988** [reply]
I am confused. Octeth is 3rd party and postal is another open source. I guess they both doesn't allow us to do anything on there code or application.  Where I should define the webhook handler? Is it separate php file in separate server besides octeth and postal server?

Also how I will get the bounce record data from postal ?

Thanks

**Reply by sinnbeck** [reply]
Yeah it can be a bit tricky. But yes a separate server with some code (a middleman so to speak). Or reach out to Octeth and ask if there is a way to handle it using their software.

I assume you are still a bit confused by the difference between an API and webhooks. With an API, ask postal for data by calling it. With webhooks postal contacts you and tells you what happened.

**Reply by kayastha1988** [reply]
Honestly, yes I am confused and trying to figure out the issue. Even octeth and postal, both doesn't have proper documentation to follow easily.
